### PR TITLE
feat(core): Introduced Duration class

### PR DESCRIPTION
**Pull request by RomainMuller**
This can be used to model durations in a user-friendly way, while
allowing coercion into the desired/required time unit at the usage site.

There is now a default requirement that properties exposed by the APIs
and that have a duration-like name (ending in `duration`, `period`,
`timeout` or `ttl`) have a type of `@aws-cdk/cdk.Duration` and do not
contain a unit suffix (`Days`, `Millis`, `Seconds`, `Sec`, ...).

BREAKING CHANGE: Properties throughout the AWS Construct Libraries that
                 represent lengths of time have been re-typed to be
                 `@aws-cdk/cdk.Duration` instead of `number`, and were
                 renamed to exclude any unit indication.

----

### Pull Request Checklist

* [x] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
  - __Design__: For significant features, design document added to `design` folder
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** and module name in parens, which will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [x] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/@aws-cdk/cdk/lib/duration.ts [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+    this.unit = unit;
+  }
+
+  /**
+   * @returns the value of this `Duration` expressed in Seconds.
+   */
+  public toSeconds(_opts: TimeConversionOptions = {}): number {
+    switch (this.unit) {
+      case TimeUnit.Seconds:
+        return this.amount;
```
Comment (rix0rrr) [inline packages/@aws-cdk/cdk/lib/duration.ts]: Feels like this repeated lookup table should not be necessary. It's too easy to make mistakes in.

If we can express every unit in terms of seconds, we can calculate from every unit to every other one.

File: packages/@aws-cdk/cdk/lib/duration.ts [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+}
+
+/**
+ * Helpful syntax sugar for turning an optional `Duration` into a count of seconds.
+ *
+ * @param duration an optional duration to convert into an amount of sunctions.
+ *
+ * @return `duration && duration.toSeconds()`.
+ */
+export function toSeconds(duration: Duration | undefined): number | undefined {
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply packages/@aws-cdk/cdk/lib/duration.ts]: I mean there's a log of places where this saves *tons* of typing, but happy to get rid of it if you feel so strongly.

File: packages/@aws-cdk/cdk/lib/duration.ts (hunk 1 above)
Comment (RomainMuller) [reply packages/@aws-cdk/cdk/lib/duration.ts]: That's a pretty good point!

File: packages/@aws-cdk/cdk/lib/duration.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (rix0rrr) [reply packages/@aws-cdk/cdk/lib/duration.ts]: Then I would like it on a class somewhere. But I'd rather get rid of it completely, or at least not put it in `@aws-cdk/cdk`.

[the remark this answers is not in the document]
Comment (RomainMuller) [reply packages/@aws-cdk/cdk/lib/duration.ts]: Sure, will remove.

File: packages/@aws-cdk/aws-apigateway/lib/stage.ts [hunk 3]
```diff
@@ -256,7 +256,7 @@ export class Stage extends Resource {
       return {
         httpMethod, resourcePath,
         cacheDataEncrypted: options.cacheDataEncrypted,
-        cacheTtlInSeconds: options.cacheTtlSeconds,
+        cacheTtlInSeconds: toSeconds(options.cacheTtl),
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply packages/@aws-cdk/aws-apigateway/lib/stage.ts]: There *is* `duration.toSeconds` - the difference is `toSeconds` is `undefined`-safe... it's basically sugar for `options.cacheTtl && options.cacheTtl.toSeconds()`. But I'll de-sugar per @rix0rrr's request.

File: packages/@aws-cdk/aws-applicationautoscaling/lib/target-tracking-scaling-policy.ts [hunk 4]
```diff
@@ -115,13 +115,6 @@ export class TargetTrackingScalingPolicy extends cdk.Construct {
       throw new Error(`Exactly one of 'customMetric' or 'predefinedMetric' must be specified.`);
     }
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply packages/@aws-cdk/aws-applicationautoscaling/lib/target-tracking-scaling-policy.ts]: Those two checks are done by the `Duration` class already. Good point about the missing `token.isUnresolved` check - I'll add that there, too.

File: packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts [hunk 5]
```diff
@@ -415,7 +411,7 @@ export class AutoScalingGroup extends AutoScalingGroupBase implements
 
     const { subnetIds } = props.vpc.selectSubnets(props.vpcSubnets);
     const asgProps: CfnAutoScalingGroupProps = {
-      cooldown: props.cooldownSeconds !== undefined ? `${props.cooldownSeconds}` : undefined,
```
[the remark this answers is not in the document]
Comment (RomainMuller) [reply packages/@aws-cdk/aws-autoscaling/lib/auto-scaling-group.ts]: Great idea!

File: packages/@aws-cdk/aws-cloudtrail/package.json [hunk 6]
```diff
@@ -88,8 +88,14 @@
     "@aws-cdk/aws-s3": "^0.35.0",
     "@aws-cdk/cdk": "^0.35.0"
   },
+  "awslint": {
+    "exclude": [
+      "duration-prop-type:@aws-cdk/aws-cloudtrail.TrailProps.cloudWatchLogsRetentionTimeDays",
```
Comment (eladb) [inline packages/@aws-cdk/aws-cloudtrail/package.json]: Should we just call this `cloudWatchLogsRetention`?

File: packages/@aws-cdk/cdk/test/test.duration.ts [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+import { Duration, Stack, Token } from '../lib';
+
+export = nodeunit.testCase({
+  'negative amount'(test: nodeunit.Test) {
+    test.throws(() => Duration.seconds(-1), /negative/);
+
+    test.done();
+  },
+
+  'unresolved amount'(test: nodeunit.Test) {
```
Comment (eladb) [inline packages/@aws-cdk/cdk/test/test.duration.ts]: Mmmm.... I wonder... If no unit conversion is needed, there shouldn't be a problem supporting tokens, and it's actually something we should support:

```ts
Duration.seconds(durationParameter.valueAsNumber).toSeconds()
```

This should work...

File: packages/@aws-cdk/cdk/lib/resource-policy.ts [hunk 8]
```diff
@@ -61,7 +63,7 @@ export interface ResourceSignal {
    * The timeout period starts after AWS CloudFormation starts creating the resource, and the timeout expires no sooner
    * than the time you specify but can occur shortly thereafter. The maximum time that you can specify is 12 hours.
    */
-  readonly timeout?: string;
+  readonly timeout?: Duration;
```
Comment (eladb) [inline packages/@aws-cdk/cdk/lib/resource-policy.ts]: Not sure we should change this... It's an "L1"

File: packages/@aws-cdk/cdk/lib/duration.ts [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+  /**
+   * @param amount the amount of Days the `Duration` will represent.
+   * @returns a new `Duration` representing `amount` Days.
+   */
+  public static days(amount: number): Duration {
+    return new Duration(amount, TimeUnit.Days);
+  }
+
+  /**
+   * @param duration an ISO-formtted duration to be parsed.
```
Comment (eladb) [inline packages/@aws-cdk/cdk/lib/duration.ts]: Link to the spec for ISO formatted durations

File: packages/@aws-cdk/cdk/lib/duration.ts [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+      if (!str) { return 0; }
+      return Number(str);
+    }
+  }
+
+  private readonly amount: number;
+  private readonly unit: TimeUnit;
+
+  private constructor(amount: number, unit: TimeUnit) {
+    if (Token.isUnresolved(amount)) {
```
Comment (eladb) [inline packages/@aws-cdk/cdk/lib/duration.ts]: Nope... We need to support number tokens as long as they don't require conversion

File: packages/@aws-cdk/cdk/lib/duration.ts [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+      case TimeUnit.Hours:
+        return `PT${this.fractionDuration('H', 24, Duration.days)}`;
+      case TimeUnit.Days:
+        return `PT${this.amount}D`;
+      default:
+        throw new Error(`Unexpected time unit: ${this.unit}`);
+    }
+  }
+
+  public toString(): string {
```
Comment (eladb) [inline packages/@aws-cdk/cdk/lib/duration.ts]: Can you add a comment explaining what this is doing. It's a cool idea...

File: packages/@aws-cdk/cdk/lib/duration.ts [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+    if (!matches) {
+      throw new Error(`Not a valid ISO duration: ${duration}`);
+    }
+    const [, days, hours, minutes, seconds] = matches;
+    if (!days && !hours && !minutes && !seconds) {
+      throw new Error(`Not a valid ISO duration: ${duration}`);
+    }
+    return Duration.seconds(
+      _toInt(seconds)
+      + (_toInt(minutes) * 60)
```
Comment (eladb) [inline packages/@aws-cdk/cdk/lib/duration.ts]: Can you reuse TimeUnit here? Seems like you have those values embedded in there..

File: packages/@aws-cdk/cdk/lib/cfn-resource.ts [hunk 13]
```diff
@@ -231,7 +231,7 @@ export class CfnResource extends CfnRefElement {
             Type: this.cfnResourceType,
             Properties: ignoreEmpty(this.cfnProperties),
             DependsOn: ignoreEmpty(renderDependsOn(this.dependsOn)),
-            CreationPolicy: capitalizePropertyNames(this, this.options.creationPolicy),
+            CreationPolicy:  capitalizePropertyNames(this, renderCreationPolicy(this.options.creationPolicy)),
```
Comment (eladb) [inline packages/@aws-cdk/cdk/lib/cfn-resource.ts]: I'd just remove this. It's not worth it.

File: packages/@aws-cdk/aws-cloudtrail/package.json (hunk 6 above)
Comment (RomainMuller) [reply packages/@aws-cdk/aws-cloudtrail/package.json]: I reckon we should!

File: packages/@aws-cdk/cdk/lib/resource-policy.ts (hunk 8 above)
Comment (RomainMuller) [reply packages/@aws-cdk/cdk/lib/resource-policy.ts]: Well yeah sure, but there is no L2 way to access this, and in any case, I feel like this is better specified in more semantic ways... I think thes may be the only use-case for the ISO-formatted duration, though...

**Comment by jayfry1077** [comment]
i'm honestly just curious as to why the decision to use Duration was enforced specifically for Lambda. 

I find it odd that when setting up a lambda i have to do 

cdk.Duration.seconds(30)  for timeout but I'm fine doing memorySize: 512.

I understand the convenience aspect but why the mandate? Why not just keep it consistent and allow a number for both or update memorySize have to use the Size class?

**Pull request merged**